e2e/cluster/cluster.go

```diff
@@ -0,0 +1,503 @@
+	return clusterAzureNetwork, clusterAzureNetworkError
+}
+
+func createNewClusterWithClient(ctx context.Context, cluster *armcontainerservice.ManagedCluster) (*Cluster, error) {
+	createdCluster, err := createNewClusterWithRetry(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	kube, err := getClusterKubeClient(ctx, config.ResourceGroupName, *cluster.Name)
+	if err != nil {
+		return nil, fmt.Errorf("get kube client using cluster %q: %w", *cluster.Name, err)
+	}
+
+	if err := ensureDebugDaemonset(ctx, kube); err != nil {
+		return nil, fmt.Errorf("unable to ensure debug damonset of viable cluster %q: %w", *cluster.Name, err)
+	}
+
+	return &Cluster{Model: createdCluster, Kube: kube}, nil
+
+}
+
+func createNewCluster(ctx context.Context, cluster *armcontainerservice.ManagedCluster) (*armcontainerservice.ManagedCluster, error) {
+	log.Printf("Creating new cluster %s in rg %s\n", *cluster.Name, *cluster.Location)
+	pollerResp, err := config.Azure.AKS.BeginCreateOrUpdate(
+		ctx,
+		config.ResourceGroupName,
+		*cluster.Name,
+		*cluster,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin aks cluster creation: %w", err)
+	}
+
+	clusterResp, err := pollerResp.PollUntilDone(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait for aks cluster creation %w", err)
+	}
+
+	return &clusterResp.ManagedCluster, nil
+}
+
+// createNewClusterWithRetry is a wrapper around createNewCluster
+// that retries creating a cluster if it fails with a 409 Conflict error
+// clusters are reused, and sometimes a cluster can be in UPDATING or DELETING state
+// simple retry should be sufficient to avoid such conflicts
+func createNewClusterWithRetry(ctx context.Context, cluster *armcontainerservice.ManagedCluster) (*armcontainerservice.ManagedCluster, error) {
+	maxRetries := 10
+	retryInterval := 30 * time.Second
+	var lastErr error
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		log.Printf("Attempt %d: Creating new cluster %s in rg %s\n", attempt+1, *cluster.Name, *cluster.Location)
+
+		createdCluster, err := createNewCluster(ctx, cluster)
+		if err == nil {
+			return createdCluster, nil
+		}
+
+		// Check if the error is a 409 Conflict
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 409 {
+			lastErr = err
+			log.Printf("Attempt %d failed with 409 Conflict: %v. Retrying in %v...\n", attempt+1, err, retryInterval)
+
+			select {
+			case <-time.After(retryInterval):
+				// Continue to next iteration
+			case <-ctx.Done():
+				return nil, fmt.Errorf("context canceled while retrying cluster creation: %w", ctx.Err())
+			}
+		} else {
+			// If it's not a 409 error, return immediately
+			return nil, fmt.Errorf("failed to create cluster: %w", err)
+		}
+	}
+
+	return nil, fmt.Errorf("failed to create cluster after %d attempts due to persistent 409 Conflict: %w", maxRetries, lastErr)
+}
+
+func getKubenetClusterModel(name string) *armcontainerservice.ManagedCluster {
```

Comment:
I'd opt for having a separate file dedicated to holding all these different mutators we use for modifying cluster models